bpf/bpf_lxc.c

```diff
@@ -463,6 +471,33 @@ static inline int handle_ipv4(struct __sk_buff *skb)
 		return DROP_POLICY;
 	}
 
+	if (ct_state_reply.proxy_port) {
+		union macaddr host_mac = HOST_IFINDEX_MAC;
+		int ret;
+
+		ret = ipv4_redirect_to_host_port(skb, &csum_off, l4_off,
+						 ct_state_reply.proxy_port, dport,
+						 orig_dip, &tuple);
+		if (IS_ERR(ret))
+			return ret;
+
+		/* After L4 write in port mapping: revalidate for direct packet access */
+		data = (void *) (long) skb->data;
+		data_end = (void *) (long) skb->data_end;
+		ip4 = data + ETH_HLEN;
+		if (data + sizeof(*ip4) + ETH_HLEN > data_end)
+			return DROP_INVALID;
+
+		cilium_trace(skb, DBG_TO_HOST, skb->cb[CB_POLICY], 0);
+
+		ret = ipv4_l3(skb, l3_off, (__u8 *) &router_mac.addr, (__u8 *) &host_mac.addr, ip4);
+		if (ret != TC_ACT_OK)
+			return ret;
+
+		cilium_trace_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
+		return redirect(HOST_IFINDEX, 0);
```

Comment:
Shouldn't the `redirect(HOST_IFINDEX, 0)` be a redirect to ingress and not egress when we send the pkt up the stack to the socket?

Could we add lib/host_port.h and place that as a new function there along with ipv4_redirect_to_host_port() itself?
Also ipv4_redirect_to_host_port() may be better named as ipv4_rewrite_to_host_port() as it better says what the function does (rewriting content)?